### PR TITLE
[SPARK-43137][SQL] Improve ArrayInsert if the position is foldable and equals to zero.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, Spark supports the `array_insert` and `array_prepend`. Users insert an element into the head of array is common operation. Considered, we want make `array_prepend` reuse the implementation of `array_insert`, but it seems a bit performance worse if the position is foldable and equals to zero.
The reason is that always do the check for position is negative or positive, and the code is too long. Too long code will lead to JIT failed.  

### Why are the changes needed?
Improve `ArrayInsert` if the position is foldable and equals to zero.

### Does this PR introduce _any_ user-facing change?
'No'.
Just change the inner implementation.

### How was this patch tested?
Exists test cases.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @infoankitp @navinvishy cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -4819,148 +4825,181 @@ case class ArrayInsert(srcArrayExpr: Expression, posExpr: Expression, itemExpr:
...
+           |  ${CodeGenerator.setArrayElement(
+                values, elementType, itemInsertionIndex, item, Some(insertedItemIsNull))}
+           |
+           |  for (int $j = $pos + $arr.numElements(); $j < 0; $j ++) {
+           |    $values.setNullAt($j + 1 + java.lang.Math.abs($pos + $arr.numElements()));
+           |  }
+           |
+           |  ${ev.value} = $values;
+           |} else {
+           |
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:4971]: can you take a close look at the generated code and see how we can simplify it if the position is a constant value? By just 10 second look, I found that if the position is constant and positive, the code can be simplified to
```
final int $resLength = java.lang.Math.max($arr.numElements() + 1, $pos);
...
```

Let's not be so specific that only target the case when `position` is `1`. We can generate code when position is constant and positive, which is a very common case.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:4971]: @cloud-fan https://github.com/apache/spark/pull/40833 used to replace this one.

**Comment by beliefer** [comment]
https://github.com/apache/spark/pull/40833 merged, so close this one.

**Pull request closed without merging**